### ISSUE TITLE
Errors with some approaches to closing windows

### ISSUE THREAD

**Issue by grahamperrin** [trueos/lumina#288]
1. launch QTerminal
2. type `exit`

Expected: closure of the window and quit of the application. 

Actual result: 

> **Application Error**
> The following application experienced an error and needed to close: 
> 
> qterminal

---

Another application that seems prone to errors at close time is the SysAdm client. For example (details taken from the error dialogue): 

```
 Open WebSocket:  URL: "wss://127.0.0.1:12150"
Open Localhost window: On Page: ""
Init UI
Connection Authorized: "127.0.0.1"
Got Server Authentication
Got Updates reply: "page_updates_checkup" "response" "sysadm" QJsonValue(object, QJsonObject({"checkupdates":{"details":"Boot-strapping updater...FAILED Fetch\nUpdating trueos-base repository catalogue...\npkg-static: http://pkg.cdn.trueos.org/master/amd64-base/meta.txz: No address record\nrepository trueos-base has no meta file, using default settings\npkg-static: http://pkg.cdn.trueos.org/master/amd64-base/packagesite.txz: No address record\nUnable to update repository trueos-base\nUpdating trueos-major repository catalogue...\npkg-static: http://pkg.cdn.trueos.org/master/amd64/meta.txz: No address record\nrepository trueos-major has no meta file, using default settings\npkg-static: http://pkg.cdn.trueos.org/master/amd64/packagesite.txz: No address record\nUnable to update repository trueos-major\nERROR: Error 70: /usr/local/sbin/pkg-static update\n\n","last_check":"2016-10-15T20:05:52","status":"noupdates"}}))
Got update check: "noupdates"
Got Updates reply: "page_updates_list_settings" "response" "sysadm" QJsonValue(object, QJsonObject({"listsettings":{"maxbe":"5","package_set":"EDGE"}}))
Got Updates reply: "page_updates_listbranch" "response" "sysadm" QJsonValue(object, QJsonObject({"listbranches":{}}))
 - Connection Closed: "127.0.0.1"
```

**Comment by beanpole135** [comment]
I am unable to reproduce either of these issues on an up-to-date TrueOS system.
In either case, this would be a problem with the application itself rather than something within Lumina.

**Comment by grahamperrin** [comment]
I've been told that posting under closed issues is frowned upon, so I posted to the lounge in Discourse but that was viewed as spam, so I'll try here and hope for no frowns … 

---- 

I can't recall which machine I used when I reported what's above but I spent some time … with the latest TrueOS Desktop on an HP EliteBook 8540p.

In Lumina, error dialogues appear when exiting from applications such as Insight, Lumina Text Editor, Lumina Screenshot, Lumina Search, Qterminal and TrueOS PDF Viewer. The 8540p is rarely touched, so I'm not bothered, but I thought I should mention it. If anyone else can reproduce the symptoms, I wonder whether there's a common underlying cause.

I tried some of the same applications with KDE, all of them with MATE: no errors when exiting.

----

Postscript: https://forums.pcbsd.org/thread-21015-post-115245.html#pid1152451 reminds me, NVIDIA … but https://www.trueos.org/downloads/1 no longer mentions known issues with NVIDIA hardware.

**Comment by grahamperrin** [comment]
Would you like to reopen this issue? Or leave it closed with an assumption that it's blocked by something upstream?

> … applications such as Insight, Lumina Text Editor, Lumina Screenshot, Lumina Search, Qterminal and TrueOS PDF Viewer. …

Other applications that have errors at close time include: 

- Lumina Archiver
- Monitor Configuration
- TrueOS Handbook (I guess, because it also uses `pc-webkitviewer`). 

Since this issue was first opened, the HP EliteBook 8540p has gained at least one fresh installation so I'm reasonably confident that these errors are not the result of 'dregs' or cruft. 

Frames from a very recent screen recording: 

![vlcsnap-2016-12-18-16h46m15s476](https://cloud.githubusercontent.com/assets/192271/21295031/ef085616-c543-11e6-83e1-2043df3f27fa.png)

![vlcsnap-2016-12-18-16h46m35s472](https://cloud.githubusercontent.com/assets/192271/21295034/fc3dafd4-c543-11e6-9001-c51c24e3cda0.png)

![vlcsnap-2016-12-18-16h47m46s144](https://cloud.githubusercontent.com/assets/192271/21295036/045c79ac-c544-11e6-9a5f-bb4abdcf83dc.png)

![vlcsnap-2016-12-18-16h47m52s340](https://cloud.githubusercontent.com/assets/192271/21295037/0951bf08-c544-11e6-8a20-cf5df98e6fdf.png)

![vlcsnap-2016-12-18-16h48m03s189](https://cloud.githubusercontent.com/assets/192271/21295040/117e1384-c544-11e6-9ecd-8a3662039f54.png)

![vlcsnap-2016-12-18-16h48m24s132](https://cloud.githubusercontent.com/assets/192271/21295047/327d2502-c544-11e6-8207-28a521d8db84.png)

![vlcsnap-2016-12-18-16h48m42s256](https://cloud.githubusercontent.com/assets/192271/21295049/3dd9d148-c544-11e6-94ef-1ece0cb8343a.png)

**Comment by q5sys** [comment]
Can you get any errors if you start these from the command line and then close them however you are that leads to the errors?
Also can you give a list of everything you've installed and are running when you get these errors.

**Comment by grahamperrin** [comment]
Aiming for a clean test environment, I created, activated and booted a new boot environment and then removed everything except: 

- pkg
- qterminal
- qupzilla-qt5
- trueos-desktop

Then `sudo truss lumina-textedit` followed by a click on the close box of the window of the application and … I won't detail what happened next (it's almost certainly off-topic from this issue) but the tab with the terminal output was lost. 

Around that time I aimed to use pkg to reinstall something, realised that the notebook was without an Internet connection (a separate issue), tried to run `rc-status` (preparing to run `rc-service` commands) but 

> Shared object "libeinfo.so.1" not found, required by …

– that was half-expected following use of AppCafe (a separate issue) and I have a hunch that the absence of that object will cause this boot environment to no longer boot so without delay I used `beadm` to activate the previous environment. I can do little more of use in this environment … 

----

… but before I left it I observed something relevant: 

- **no on-screen application error** when the window of QTerminal closes in response to me entering `exit` in its last tab.

**Comment by grahamperrin** [comment]
> … https://www.trueos.org/downloads/ no longer mentions known issues with NVIDIA hardware.

I encountered the symptom (presumably the same bug) with **QupZilla on a fresh installation** of TrueOS Desktop on a MacBookPro8,2. From https://forums.pcbsd.org/thread-20098-post-111116.html#pid111116:

>> Intel Core i7 (2635QM, 2720QM, 2820QM) ("Sandy Bridge")

>> AMD Radeon HD 6490M or 6750M and Intel HD Graphics 3000

Speeding through `/var/log/Xorg.0.log` I think that the OS uses SCFB in this case.

**Comment by baitisj** [comment]
I'm unable to reproduce this issue, too. I wonder if you have some strange qt preference somewhere that is precipitating the error

**Comment by grahamperrin** [comment]
> strange qt preference

Unlikely. The installation on the MacBookPro8,2 is not only almost completely clean (nothing other than QupZilla added), it's also very young – only a few hours old. I performed the installation for the sole purpose of reviewing this issue, I have barely touched it. 

I can't easily provide the details from the most recent error dialogue, I'll get to that later.

**Comment by baitisj** [comment]
If you create a new BE, I think that it doesn't change how your user directories are mounted. On my system, the base system and the user directories are completely different datasets.

Try creating a new *user* and log in as that person, and see if you can reproduce.

**Comment by grahamperrin** [comment]
> a new BE

Already more than that; it's a fresh installation. (Sorry, force of habit, in my previous comment I wrote clean when I meant fresh. I'll correct the comment.)

**Comment by baitisj** [comment]
It's a fresh installation: but...
* have you done any updates?
* are you on EDGE or UNSTABLE?
( paste output of `pkg version | grep trueos-desktop` )

**Comment by grahamperrin** [comment]
> I can't easily provide the details …

Mentioned in Gitter, it's the machine with a terribly slow network connection. Pasting is excruciating. Respecting wishes for not too many notifications, let's continue in the troubleshooting room

**Comment by baitisj** [comment]
I attempted to reproduce this in VirtualBox with the most recent STABLE release, as well as my machine (UNSTABLE) -- which has NVidia hardware, and have not been able to reproduce the issue.

**Comment by grahamperrin** [comment]
```
$ date ; uptime ; uname -v ; uname -n
Fri 23 Dec 2016 05:19:14 GMT
 5:19a.m.  up 28 mins, 0 users, load averages: 1.00, 0.96, 0.83
FreeBSD 12.0-CURRENT #10 a4012d9(drm-next-4.7): Fri Dec 23 00:27:25 UTC 2016     root@gauntlet:/usr/obj/usr/src/sys/GENERIC 
momh167-gjp4-macbookpro112-trueos.university.brighton.ac.uk
$ 
```

I have an _Application Error_ dialogue for `qsudo pc-diskmanager` after I decline to enter the password for Disk Manager. Details: 

> sudo: a password is required

True: a password is required, but simply declining the invitation should not lead to a modal dialogue with a red alert for an error. 

-----

I'll continue to test at this Mac, with … all things considered, I'll focus on Lumina Text Editor.

**Comment by grahamperrin** [comment]
https://drive.google.com/file/d/0B1G6-Ym42bEidXNJcE91Z0ZEakk/view

**Comment by grahamperrin** [comment]
From around seven weeks ago: 

[2016-11-03 04-06 crash of Qterminal when I commanded exit.txt](https://github.com/trueos/lumina/files/671642/2016-11-03.04-06.crash.of.Qterminal.when.I.commanded.exit.txt)

**Comment by grahamperrin** [comment]
> I can't find the output from truss. Sorry.

The required file was misfiled as a result of Lumina bug 336. Now found: 

- [2016-12-23 06-09 -ish SIGNAL 10 (SIGBUS) with a sudo truss run of lumina-search.txt](https://github.com/trueos/lumina/files/671647/2016-12-23.06-09.-ish.SIGNAL.10.SIGBUS.with.a.sudo.truss.run.of.lumina-search.txt)

**Comment by grahamperrin** [comment]
From https://gitter.im/trueos/Random?at=585e10900730ce6937044b97: 

> … output file (from truss) provides the cause/background of the error 

Maybe. If not, I hope that the output contains enough for a developer to walk me through a more appropriate approach to debugging this issue. 

> but truss itself is not what causes the error, correct?

That's probably true. 

----

Years ago I began lightweight use of [dtruss](https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man1/dtruss.1m.html)(1m) not as a developer, but as a tester with AppleSeed projects. Some time after shifting from Mac OS X to TrueOS, thanks to @q5sys at https://discourse.trueos.org/t/-/124/3?u=grahamperrin in November I learnt of [truss](https://www.freebsd.org/cgi/man.cgi?query=truss&sektion=1&manpath=FreeBSD)(1) and remained aware that simple `sudo`/root use of _dtruss_ was **sometimes undesirable** so three weeks ago, I dug into Stack Exchange and found things such as: 

- [How to trace a program from its very beginning without running it as root](http://stackoverflow.com/q/1204256/38108) (2009-07-30)
- [How can get dtrace to run the traced command with non-root privileges?](http://stackoverflow.com/q/3007868/38108) (2010-06-09)

A few days ago I discovered that `truss` does not require superuser privileges but old `dtruss` habits die hard, and the name of the file in my previous comment –

> 2016-12-23 06-09 -ish SIGNAL 10 (SIGBUS) with a sudo truss run of lumina-search.txt

– suggests that I used `sudo` when possibly, I did **not**. (Whenever I work as root in a terminal where output might be saved, I habitually run `whoami`; no such run in yesterday's output so I assume that I worked as a different user.)

**Comment by beanpole135** [comment]
See if this makes it work better for you[1].
I disabled the checking of the return code, and only look for actual process crashes now, since it is clear that some apps return non-zero even when closing successfully.

[1] https://github.com/trueos/lumina/commit/64012116a826c52c7c11e94a13309f246e3f5516

**Comment by grahamperrin** [comment]
Fix confirmed, thanks. 

I guess that the bug was limited to some combinations of graphics hardware and software.
